File: 2024/Day 2/day2.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>
#include <string>
#include <algorithm>
#include <cmath>

bool isSafe(const std::vector<int>& levels) {
    bool isIncreasing = true;
    bool isDecreasing = true;

    for (size_t i = 1; i < levels.size(); ++i) {
        int diff = levels[i] - levels[i - 1];
        if (std::abs(diff) < 1 || std::abs(diff) > 3) {
            return false;
        }
        if (diff > 0) {
            isDecreasing = false;
        }
        if (diff < 0) {
            isIncreasing = false;
        }
    }

    return isIncreasing || isDecreasing;
}

int countSafeReports(const std::string& filename) {
    std::ifstream file(filename);
    std::string line;
    int countSafe = 0;

    while (std::getline(file, line)) {
        std::istringstream iss(line);
        std::vector<int> levels((std::istream_iterator<int>(iss)), std::istream_iterator<int>());
        if (isSafe(levels)) {
            ++countSafe;
        }
    }

    return countSafe;
}

int countSafeReportsWithDampener(const std::string& filename) {
    std::ifstream file(filename);
    std::string line;
    int countSafe = 0;

    while (std::getline(file, line)) {
        std::istringstream iss(line);
        std::vector<int> levels((std::istream_iterator<int>(iss)), std::istream_iterator<int>());

        if (isSafe(levels)) {
            ++countSafe;
            continue;
        }

        for (size_t i = 0; i < levels.size(); ++i) {
            std::vector<int> modifiedLevels = levels;
            modifiedLevels.erase(modifiedLevels.begin() + i);
            if (isSafe(modifiedLevels)) {
                ++countSafe;
                break;
            }
        }
    }

    return countSafe;
}

int main() {
    std::string filename = "input.txt";

    int part1Result = countSafeReports(filename);
    int part2Result = countSafeReportsWithDampener(filename);

    std::cout << "Part 1: Number of safe reports: " << part1Result << std::endl;
    std::cout << "Part 2: Number of safe reports (with Problem Dampener): " << part2Result << std::endl;

    return 0;
}

File: 2024/Day 2/day2.py
def is_safe(levels):
    is_increasing = True
    is_decreasing = True

    for i in range(1, len(levels)):
        diff = levels[i] - levels[i - 1]
        if not (1 <= diff <= 3 or -3 <= diff <= -1):
            return False
        if diff > 0:
            is_decreasing = False
        if diff < 0:
            is_increasing = False

    return is_increasing or is_decreasing


def count_safe_reports(data):
    reports = data.strip().split('\n')
    count_safe = 0

    for report in reports:
        levels = list(map(int, report.strip().split()))
        if is_safe(levels):
            count_safe += 1

    return count_safe


def count_safe_reports_with_dampener(data):
    reports = data.strip().split('\n')
    count_safe = 0

    for report in reports:
        levels = list(map(int, report.strip().split()))
        if is_safe(levels):
            count_safe += 1
            continue
        for i in range(len(levels)):
            modified_levels = levels[:i] + levels[i + 1:]
            if is_safe(modified_levels):
                count_safe += 1
                break

    return count_safe


# Example, copy and paste input into here haha
data = """7 6 4 2 1
1 2 7 8 9
9 7 6 2 1
1 3 2 4 5
8 6 4 4 1
1 3 6 7 9"""

part1_result = count_safe_reports(data)
part2_result = count_safe_reports_with_dampener(data)

print(f"Part 1: Number of safe reports: {part1_result}")
print(f"Part 2: Number of safe reports (with Problem Dampener): {part2_result}")
